### ISSUE TITLE
Dynamic target list / Regex in target configuration

### ISSUE THREAD

**Issue by bkez322** [promregator/promregator#35]
I haven’t tried using promregator quite yet but it looks very promising and is very well documented- good work to the people behind it.

In my use case, I’d like to be able to have promregator scrape from all apps running on CF without having to manually update the config and restart promregator every time there’s a new app on CF. I don’t believe this use case is described in the documentation.

Does promregator reload the config periodically. If so, one way this can be accomplished is to have anindependent side process doing service discovery on what appnames exist across all orgs and space, and then write them to the promregator config file- similar to how Prometheus handles its file-based service discovery.

Another solution may be to have the promregator config support regex for the targets. That way I can match against all orgs/spaces/apps. This would force Promregator to periodically check all existing orgs/spaces/apps, and so this may or may not impact performance somewhat.

If there’s interest in having a built in solution to this, like using regex, and if there’s an idea in how it should be implemented, I’d consider PRing it.

**Comment by eaglerainbow** [comment]
Thank you for the praise 😄 

> I don’t believe this use case is described in the documentation.

No, it's not. It's more the other way around, that in most cases, people wanted to explicitly say, which apps they want to have scraped, because metrics support isn't build-in everywhere. But I can understand your requirement quite well, too. 

> Does promregator reload the config periodically.

So far, it does not. I did not check if spring-boot supports this, but I have my doubts on this, as it would mean that the entire application boot sequence would have to be re-executed. I get the general benefit, though.

> to have anindependent side process

Well, promregator is already such a side-car to Prometheus (and this is done more due to bare necessity than conviction). I don't think that yet-another-side-car to the side-car is something that we really want to have.

> Another solution may be to have the promregator config support regex for the targets.

That I do understand and would also support from an architectural / design point of view. Yet, it does not allow to change all attributes (there could other attributes which could change over time, too - I have the password on my mind, for example). However, it could certainly cover your case, where you want to discover targets automatically (and that should be the 80% case from my point of view). 

> If there’s interest in having a built in solution to this, like using regex,

Certainly, there is! 

> if there’s an idea in how it should be implemented

Well, I do not have the silver-bullet either, but these are my ideas, which just went through my head on this:

* The class to go to is https://github.com/promregator/promregator/blob/a31c9278b973d0b3382548f76127cf3ad2853605/src/main/java/org/cloudfoundry/promregator/scanner/ReactiveAppInstanceScanner.java
* Mainly this is about the methods `getOrgId`, `getSpaceId`, and `getApplicationId` there. Currently they only support specifying a single string. The methods need to be enhanced / overloaded / changed /..., such that also a regex is supported.
* Currently, those methods only support a `Mono<String>` as return type. That's most likely too small. However, a `Flux<String>` isn't sufficient either, as you have to be able to map the space's name to the spaceId. Most likely this is the hardest nut to crack somehow (If you are not familiar with the reactor concept (Mono/Flux), you might want to have a look at https://projectreactor.io/ ).
* Multi-org/-space/-application support is already available via the caching maps - see also  https://github.com/promregator/promregator/blob/a31c9278b973d0b3382548f76127cf3ad2853605/src/main/java/org/cloudfoundry/promregator/scanner/ReactiveAppInstanceScanner.java#L31 . Alas, the most important thing is to get the mappings into the maps.
* Caching the answers of CF in Promregator is quite important (more than you might think initially). And it also could help here to match the regex'es against the keyset of the caching maps without asking CF each time.
* Cache timeouts become even more important in your case, yet (see for example https://github.com/promregator/promregator/blob/a31c9278b973d0b3382548f76127cf3ad2853605/src/main/java/org/cloudfoundry/promregator/scanner/ReactiveAppInstanceScanner.java#L86 ). Otherwise, you would add a space, but promregator wouldn't see it for (by default) up to an hour. Most likely, a recommendation should be added to the documentation stating that the corresponding configuration option (e.g. `cf.cache.timeout.space` and `cf.cache.timeout.org`) should be adjusted to a reasonably lower value (I thought about auto-adjusting it, but I disliked the idea).
* To stay compatible with it, I'd suggest adding yet three more options to the targets in the configuration (see also https://github.com/promregator/promregator/blob/a31c9278b973d0b3382548f76127cf3ad2853605/src/main/java/org/cloudfoundry/promregator/config/Target.java ), which hold the regex - something like `orgRegex`, `spaceRegex`, `applicationRegex` or so. `orgRegex` should be mutual exclusive to `orgName` - but one of them has to be filled. We can also say that if you don't fill them, then we assume `orgRegex=^.*$` (similar for `spaceRegex` and `applicationRegex`), but that isn't without danger - think about if someone did completely miss to specify any target on a very large CF account...
* Documentation then needs to be updated accordingly, esp. docs/config.md. It might also be worthwhile to meantion this feature on the frontpage at README.md. 
* Unit tests needs to be adjusted / enhanced, especially https://github.com/promregator/promregator/blob/a31c9278b973d0b3382548f76127cf3ad2853605/src/test/java/org/cloudfoundry/promregator/scanner/ReactiveAppInstanceScannerTest.java will be affected. If things get complicated / large, It's not forbidden to create yet another unit class (e.g. junit test case).

> I’d consider PRing it.

Then I just say: Welcome on board 😃 
I'd really looking forward to review your PR!

**Comment by eloo** [comment]
Hello,
i've also tested promregator for my setup and it looks very promising.
Thus regex for target, especially for application discovery would be very nice for beginning

**Comment by tzolov** [comment]
Hi @eaglerainbow , 
For tools such as [Spring Cloud Data Flow](https://cloud.spring.io/spring-cloud-dataflow/) (SCDF) that can deploy, manage and orchestrate numerous apps, the ability to handle dynamic targets is critical. 

One possible options is for `promregator` to expose an REST point (e.g. something like POST /targets)  for updating the targets. Then SCDF can use this entry point to update the targets

A more generic options (as it has been suggested above) is to allow promregator `discover` all apps in org/space and pre-filter using some name-filter criteria. The [PR#46](https://github.com/promregator/promregator/pull/46) implements a more pragmatic version of this idea. It introduce the `SPACE_ALL_APPLICATIONS` name convention to instruct promregator to include all CF apps of the provided Org/Space.

**Comment by eaglerainbow** [comment]
@tzolov Thanks for your suggested contribution - I appreciate your ideas and work greatly (esp. as I haven't set up all the contribution-enabling items, such as CONTRIBUTION.md, Coding conventions, etc. of this project, yet)! I am currently looking into it (only have little time today, but let's see how far we get)...

> For tools such as Spring Cloud Data Flow (SCDF) [...] the ability to handle dynamic targets is critical.

Point taken - the idea of auto-discovering of apps registered on the platform and which is depicted in this issue is undisputed.

> One possible options is for promregator to expose an REST point (e.g. something like POST /targets) for updating the targets. Then SCDF can use this entry point to update the targets

Yes, that would be an option. But especially in the environment of Prometheus, the matter of discovery (here: of CF) is a pivotal topic. Moreover, in the process of developing promregator I learned that in case of the Cloud Foundry API, discovery and scraping need to go hand in hand. Letting another consumer update the configuration via a REST endpoint would mean that this application has to "re-invent the discovery" mechanism provided by promregator. I don't think that this is the right way to go. 

> [PR #46] introduce the `SPACE_ALL_APPLICATIONS` name convention to instruct promregator to include all CF apps of the provided Org/Space.

Mmmh... In the past I had some bad experience with "magic identifiers", which then triggers a special behaviour. Moreover, the suggested identifier `SPACE_ALL_APPLICATIONS` theoretically could also be a valid app name (not that I assume that this would be used very often, though). Instead, perhaps an identifier using symbols only, such as `*` or `.*` (which would bring us again more into the direction of regular expressions) would be a better option.
Moreover, I am currently thinking about it, whether making (at least) `applicationName` (which I think would be the biggest painpoint from a feature's perspective) optional and then doing the same thing as you did with `SPACE_ALL_APPLICATIONS`, couldn't also be a way to go. 

Using the magic identifier `SPACE_ALL_APPLICATIONS` will cause us to keep compatibility for it almost until eternity - even if we may have `applicationRegex` (see above) some time in the future. 

On PR #46 itself, I will have a look into it and post additional comments there when reviewing it.

**Comment by eaglerainbow** [comment]
@bkez322 Could you please have a look at PRs #46 and #47?
If merged, this could provided the two features:
* "scan all applications in a space" (#46) and 
* "scan all applications based on a regex" (#47). 

We don't have regex filter (or "get all") on space level and org level yet.
How important would this be for you? If it is important, could you give us a little insight on your use case to understand why this is so?
Background to my question: I can easily understand that you may have many new apps in a space and don't want to restart Promregator. But I struggle a bit based on my experience that the number of orgs/spaces are getting "uncontrollable"...
Thanks!

**Comment by bkez322** [comment]
@eaglerainbow Scanning all apps is certainly far more essential than orgs or spaces for the reasons already said. In the workplace environment I'd be using Promregator in, about 1-3 new CF orgs are created or renamed each week. It would *nice* from my perspective if Promregator wasn't yet another thing here that required a configuration update and/or restart every time that happens, but again, it's certainly not critical to its usability.

Perhaps my use-case is an unusual one, but I'd be using Promregator to monitor 100+ CF orgs. Again it would be *nice* if I didn't have to add all 100+ of those to a configuration file.

**Comment by eaglerainbow** [comment]
Understood and point taken - you are the customer(s); but no promise made, though 😀 
100+ CF orgs! 😮 Hell, that is really a lot... I never had thought about such a huge number of apps, when I had designed Promregator. For sure you have to use the Single Target Scraping mode for such a huge number of apps. Please expect that Promregator will have performance issues in such a configuration, as 

1. simply - as said - I never considered using it with so many apps, and 
2. I don't have access to such an environment where I could test it. 

Yet, I would be eager to hear from your experience with Promregator in such a challenging environment! Please feel free to create issues, if you can spot the cases, which don't scale properly.

**Comment by eaglerainbow** [comment]
@eloo @tzolov @bkez322 After a long battle with this feature, I think I finally made it happen that we can
* scrape all apps in a space (by leaving off `applicationName` completely), and/or
* scrape a set of apps specified by a regular expression (in `applicationRegex`).
Also proper resolution caching including cleanup of orphaned Promregator's own metrics are provided. 

The stuff is currently merged with `master`.

Given the size of this enhancement. I would be looking for volunteers (and you are the main "suspects" 😀 ), who would be testing it at least once  before I release it publicly. 

Are you interested in such an activity?

**Comment by eloo** [comment]
@eaglerainbow 
Thanks, very cool. We will try the feature this or next week 👍 
Maybe you can create and upload an dockerimage with a specific tag for this feature? So it would be easier to test ;)

**Comment by eaglerainbow** [comment]
@eloo Thanks! I'll create a release candidate for you - hopefully even today night. 
Once I have the version, I'll note down the version identifier here...

**Comment by eaglerainbow** [comment]
Here we go: The number you have to dial is `0.4.0-rc1` 😀

**Comment by eloo** [comment]
@eaglerainbow 
thanks.. i hope we can test it tomorrow ;)

**Comment by dschroe** [comment]
@eaglerainbow 
Sorry for the late response. We've tested version `0.4.0-rc2` and the new regex feature but there is a problem. We used the following (regex-)configuration:

```
cf:
  ...
  targets:
    - orgName: anyName
      spaceName: anyName
      path: /.../prometheus-metrics
      applicationRegex: ^(?=abc).*|^(?=def).*|^(?=uia).*
```
This regex pattern should be valid and should match every cf-app whose name starts either with "abc", "def" or "uia" (positive lookahead). In most cases (when invoking the /discovery or the /metrics endpoint) this regex works perfectly. But sometimes the following exception will be produced (and Promregator cannot recover from this exception):

```
2018-06-06 16:38:12.727 ERROR 17930 --- [     parallel-5] o.c.p.cfaccessor.ReactiveCFAccessorImpl  : Retrieval of retrieveAllApplicationIdsInSpace with key 257e451a-1b28-432d-b5c5-f84a90c5be25|971071fb-41d8-4a0a-9d1a-009e16c17642 raised a reactor error

java.util.concurrent.TimeoutException: Did not observe any item or terminal signal within 2500ms (and no fallback has been configured)
	at reactor.core.publisher.FluxTimeout$TimeoutMainSubscriber.handleTimeout(FluxTimeout.java:268) ~[reactor-core-3.1.7.RELEASE.jar:3.1.7.RELEASE]
	at reactor.core.publisher.FluxTimeout$TimeoutMainSubscriber.doTimeout(FluxTimeout.java:253) ~[reactor-core-3.1.7.RELEASE.jar:3.1.7.RELEASE]
	at reactor.core.publisher.FluxTimeout$TimeoutTimeoutSubscriber.onNext(FluxTimeout.java:370) ~[reactor-core-3.1.7.RELEASE.jar:3.1.7.RELEASE]
	at reactor.core.publisher.StrictSubscriber.onNext(StrictSubscriber.java:89) ~[reactor-core-3.1.7.RELEASE.jar:3.1.7.RELEASE]
	at reactor.core.publisher.FluxOnErrorResume$ResumeSubscriber.onNext(FluxOnErrorResume.java:73) ~[reactor-core-3.1.7.RELEASE.jar:3.1.7.RELEASE]
	at reactor.core.publisher.MonoDelay$MonoDelayRunnable.run(MonoDelay.java:117) ~[reactor-core-3.1.7.RELEASE.jar:3.1.7.RELEASE]
	at reactor.core.scheduler.SchedulerTask.call(SchedulerTask.java:50) ~[reactor-core-3.1.7.RELEASE.jar:3.1.7.RELEASE]
	at reactor.core.scheduler.SchedulerTask.call(SchedulerTask.java:27) ~[reactor-core-3.1.7.RELEASE.jar:3.1.7.RELEASE]
	at java.util.concurrent.FutureTask.run(FutureTask.java:266) ~[na:1.8.0_152]
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201(ScheduledThreadPoolExecutor.java:180) ~[na:1.8.0_152]
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:293) ~[na:1.8.0_152]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) ~[na:1.8.0_152]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) ~[na:1.8.0_152]
	at java.lang.Thread.run(Thread.java:748) ~[na:1.8.0_152]
```

**Comment by eaglerainbow** [comment]
> Update: It seems that this problem does not necessarily have to do something with the regex-feature. When omitting the regex and the applicationName (so that Promregator searches for all apps in the specified cf-space) the same problem occurs sometimes.

That exactly sounds like the behavior you should get, after my enhancements described in https://github.com/promregator/promregator/issues/59#issuecomment-393453513. So, I wouldn't associate that to regex scanning either.

There is one small thing, though, which draws my attention to:  

> When omitting the regex and the applicationName (so that Promregator searches for all apps in the specified cf-space) the same problem occurs sometimes.

Does this mean that you cannot reproduce the issue, if you do not omit the `applicationName`? If yes, then this would be a first hint...

BTW: I would suggest to continue the discussion in #59 on this specific topic to keep this thread here clean (and reserved for posts on the regex topic).

**Comment by eloo** [comment]
@eaglerainbow 
i have now a setup with application regex. for now it looks pretty good.

**Comment by eaglerainbow** [comment]
I finally managed to rewrite the `ReactiveTargetResolver` and implemented regular expressions also for spaces and organizations. Please have a look at the configuration options `cf.targets[].orgRegex` and `cf.targets[].spaceRegex`. 
Resolving them can be quite tricky and may require multiple requests to the CF Cloud Controller. So, specifying the `orgName` and `spaceName` is recommended, if you know them anyway (for Promregator, this makes life much easier).

The new feature is available in `0.5.0-rc6`, which I just released. 

@eloo @bkez322 @ghost @tzolov 
I would be interested in hearing your feedback on this state of implementation.

**Comment by eaglerainbow** [comment]
[0.5.0](https://github.com/promregator/promregator/releases/tag/v0.5.0) is out now.
